fabtests/hpcs/src/core/core.c

```diff
@@ -0,0 +1,1476 @@
+
+	ret = fi_close(&domain_state->domain->fid);
+	if (ret) {
+		hpcs_error("unable to close domain\n");
+		return ret;
+	}
+	domain_state->domain = NULL;
+
+	return 0;
+}
+
+/*
+ * Finalizes OFI.  Safely cleans up resources created by init_ofi.  Takes a
+ * "best effort" approach - goes as far as it can and returns the first error
+ * encountered.
+ */
+int fini_ofi(struct ofi_state *ofi_state)
+{
+	int ret;
+
+	ret = fini_domain(ofi_state->domain_state);
+	if (ret)
+		return ret;
+
+	ofi_state->domain_state = NULL;
+
+	ret = fi_close(&ofi_state->fabric->fid);
+	if (ret) {
+		hpcs_error("unable to close fabric\n");
+		return ret;
+	}
+
+	return 0;
+}
+
+/*
+ * HPCS uses groups of arguments separated by a "-" or "--" (which are treated
+ * as special by getopt, and cause it to stop parsing arguments).
+ *
+ * This function updates argc and argv to start at the next group of args,
+ * with argv[0] pointing to the separator.
+ */
+static void next_args(int *argc, char * const**argv)
+{
+	int i;
+	static char* empty = "--";
+
+	/* 0th element is either binary name or previous separator. */
+	for (i=1; i<*argc; i++) {
+		if (strcmp((*argv)[i], "-") == 0 || strcmp((*argv)[i], "--") == 0)
+			break;
+	}
+
+	if (i < *argc) {
+		*argc = *argc - i;
+		*argv = &(*argv)[i];
+	} else {
+		*argc = 1;
+		*argv = &empty;
+	}
+
+	/* Current option index global variable must be reset. */
+	optind = 1;
+}
+
+static int parse_arguments(int argc, char * const* argv,
+		struct arguments **arguments)
+{
+	int longopt_idx, op, ret = 0;
+	struct option longopt[] = {
+		{"prov", required_argument, 0, 'p'},
+		{"window", required_argument, 0, 'w'},
+		{"order", required_argument, 0, 'o'},
+		{"pattern", required_argument, 0, 'a'},
+		{"iterations", required_argument, 0, 'n'},
+		{"verbose", no_argument, &verbose, 1},
+		{"help", no_argument, 0, 'h'},
+		{0}
+	};
+
+	struct arguments *args = calloc(sizeof (struct arguments), 1);
+	int have_pattern = 0;
+
+	if (args == NULL)
+		return -FI_ENOMEM;
+
+	*args = (struct arguments) {
+		.prov_name = NULL,
+		.window_size = DEFAULT_WINDOW_SIZE,
+		.callback_order = DEFAULT_CALLBACK_ORDER,
+		.pattern_api = {0},
+		.iterations = 1
+	};
+
+	while ((op = getopt_long(argc, argv, "vp:w:o:a:n:h", longopt, &longopt_idx)) != -1) {
+		switch (op) {
+		case 0:
+			if (longopt[longopt_idx].flag != 0)
+				printf("verbose mode enabled\n");
+			break;
+		case 'p':
+			args->prov_name = calloc(1, 128);
+			if (args->prov_name == NULL)
+				return -FI_ENOMEM;
+			if (sscanf(optarg, "%127s", args->prov_name) != 1)
+				return -FI_EINVAL;
+			break;
+		case 'w':
+			if (sscanf(optarg, "%zu", &args->window_size) != 1)
+				return -FI_EINVAL;
+			break;
+		case 'o':
+			if (strcmp(optarg, "none") == 0)
+				args->callback_order = CALLBACK_ORDER_NONE;
+			else if (strcmp(optarg, "expected") == 0)
+				args->callback_order = CALLBACK_ORDER_EXPECTED;
+			else if (strcmp(optarg, "unexpected") == 0)
+				args->callback_order = CALLBACK_ORDER_UNEXPECTED;
```

Comment:
I don't understand why this is a user facing option

Reply:
Our internal name is bad, since it doesn't really pertain to callbacks.  This is how the user selects between expected messaging or unexpected messaging, by forcing either the send or the receive to be posted first and doing an MPI barrier in the middle.  Order "none" means you don't care, and there's no barrier.

(We added a "--repeat=N" option to the onesided test to get around the performance impact of the barrier; you can set "--order=expected" and a large number of repeats in order to do a burst of RMA writes or whatever that are only blocked by a single barrier at the beginning.  If memory serves, I think "--order=none" works with one-sided RMA now, so maybe it's not as important as it was.  However, it might be nice to have something like that for the "sendrecv" test, for cases where you want to stress-test only unexpected messages or only expected messages.)